### PR TITLE
Solves #15 Safe files to partial submissions

### PR DESCRIPTION
**Pull request by Firesphere**
- Submit the file(s) through Ajax
- Attempt to save as a partial submission

### CODE REVIEW

File: tests/unit/PartialFileFieldSubmissionTest.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    protected function setUp()
+    {
+        $this->field = PartialFileFieldSubmission::create();
+        $partialForm = PartialFormSubmission::create();
+        $udf = UserDefinedForm::create(['Title' => 'Test'])->write();
+        $partialForm->UserDefinedFormID = $udf;
+        $partialFormID = $partialForm->write();
+        $this->field->SubmittedFormID = $partialFormID;
+
+        return parent::setUp();
```
Comment (lhalaa) [inline tests/unit/PartialFileFieldSubmissionTest.php]: Can you move the `parent::setUp()` at the beginning of this function and remove `return`? This also stops the `Call to a member function activate() on boolean` error.

File: tests/unit/PartialSubmissionControllerTest.php [hunk 2]
```diff
@@ -53,6 +56,23 @@ public function testSavePartialSubmissionFieldCreated()
         $this->assertCount(1, $fields);
     }
 
+    public function testSavePartialSubmissionFileField()
+    {
+        $form = $this->objFromFixture(UserDefinedForm::class, 'form1');
```
Comment (lhalaa) [inline tests/unit/PartialSubmissionControllerTest.php]: This form is already published in the `setUp()`

File: tests/unit/PartialSubmissionControllerTest.php [hunk 3]
```diff
@@ -53,6 +56,23 @@ public function testSavePartialSubmissionFieldCreated()
...
+        $form->publishRecursive();
+        $field = $this->objFromFixture(EditableFileField::class, 'filefield1');
+        $field->write();
+        $multipart = [
+            'name'     => 'file1',
+            'contents' => fopen(__DIR__ . '/../fixtures/Hans-fullsize-sqr.png', 'rb'),
+            'filename' => 'Hans-fullsize-sqr.png'
+        ];
+
+        $response = Director::test('/partialuserform/save', ['File' => $multipart], [], 'POST');
```
Comment (lhalaa) [inline tests/unit/PartialSubmissionControllerTest.php]: Any reason why you are not using `$this->savePartial()` or `$this->post()`?

File: src/extensions/UserDefinedFormControllerExtension.php [hunk 4]
```diff
@@ -2,19 +2,39 @@
...
+     * @var null|PartialFormSubmission
+     */
+    protected $partialForm;
+
     public function onBeforeInit()
     {
         Requirements::javascript('firesphere/partialuserforms:client/dist/main.js');
     }
+
+    public function verify()
```
Comment (lhalaa) [inline src/extensions/UserDefinedFormControllerExtension.php]: I'm assuming you're planning to use this in the future?

File: src/extensions/UserDefinedFormControllerExtension.php [hunk 5]
```diff
@@ -2,19 +2,39 @@
...
+    {
+        return $this;
+    }
+
+    public function getPartialForm()
```
Comment (lhalaa) [inline src/extensions/UserDefinedFormControllerExtension.php]: I'm assuming you're planning to use this in the future?

File: tests/unit/PartialFileFieldSubmissionTest.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+use Firesphere\PartialUserforms\Models\PartialFileFieldSubmission;
+use Firesphere\PartialUserforms\Models\PartialFormSubmission;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Security\DefaultAdminService;
+use SilverStripe\Security\Security;
+use SilverStripe\UserForms\Model\UserDefinedForm;
+
+class PartialFileFieldSubmissionTest extends SapphireTest
+{
+
```
Comment (lhalaa) [inline tests/unit/PartialFileFieldSubmissionTest.php:14]: Can you please add `protected $usesDatabase = true;` so it uses tempdb and doesn't insert test data into the local database?

**Review by lhalaa — COMMENTED** [review]
minor stuff @Firesphere, also tests are failing on my local 🤷‍♀

File: tests/unit/PartialSubmissionControllerTest.php (hunk 3 above)
Comment (Firesphere) [reply tests/unit/PartialSubmissionControllerTest.php]: Pure habit

File: src/extensions/UserDefinedFormControllerExtension.php (hunk 4 above)
Comment (Firesphere) [reply src/extensions/UserDefinedFormControllerExtension.php]: Nope, should not have been committed on this branch 🤦‍♂

File: src/extensions/UserDefinedFormControllerExtension.php (hunk 5 above)
Comment (Firesphere) [reply src/extensions/UserDefinedFormControllerExtension.php]: As above 🤦‍♂

File: tests/unit/PartialFileFieldSubmissionTest.php (hunk 1 above)
Comment (Firesphere) [reply tests/unit/PartialFileFieldSubmissionTest.php]: Yeps, habit :)

File: tests/unit/PartialSubmissionControllerTest.php (hunk 2 above)
Comment (Firesphere) [reply tests/unit/PartialSubmissionControllerTest.php]: D'oh :D

File: src/models/PartialFormSubmission.php [hunk 7]
```diff
@@ -52,25 +58,39 @@ class PartialFormSubmission extends SubmittedForm
...
         $fields = parent::getCMSFields();
-        $fields->removeByName(['Values', 'IsSend', 'PartialFields', 'TokenSalt', 'Token', 'UserDefinedFormID', 'Submitter']);
+        $fields->removeByName([
+            'Values',
+            'IsSend',
+            'PartialFields',
+            'TokenSalt',
+            'Token',
+            'UserDefinedFormID',
+            'Submitter'
```
Comment (lhalaa) [inline src/models/PartialFormSubmission.php]: Do you want to remove `PartialUploads` as well, since it's merged into `PartialFields`?

**Review by lhalaa — CHANGES_REQUESTED** [review]
Doesn't work on my machine. I can't see the uploaded file in the partial page and CMS

File: tests/unit/PartialUserFormControllerTest.php [hunk 8]
```diff
@@ -12,9 +12,6 @@
  */
 class PartialUserFormControllerTest extends FunctionalTest
 {
-    /**
-     * @var string
-     */
```
Comment (Firesphere) [inline tests/unit/PartialUserFormControllerTest.php]: ... Shouldn't have been removed?

Comment (lhalaa) [reply tests/unit/PartialUserFormControllerTest.php]: 🤷‍♀

File: src/models/PartialFormSubmission.php (hunk 7 above)
Comment (Firesphere) [reply src/models/PartialFormSubmission.php]: Yeps, good thing to make sure it's not duplicate (although I didn't see it, better safe than sorry)

**Comment by Firesphere** [comment]
There was a rebase-error that caused it to be broken. Fixed it now again :)

File: client/src/js/partialuserforms/partialsubmission.js [hunk 9]
```diff
@@ -22,9 +22,13 @@ const getElementValue = (element, fieldName) => {
       return valueArray;
     }
     return '';
+  } else if (element.getAttribute('type') === 'file') {
+    return element.files[0];
   } else {
     return value;
   }
+
+  return value;
```
Comment (lhalaa) [inline client/src/js/partialuserforms/partialsubmission.js:34]: Please remove this line

File: client/src/js/partialuserforms/partialsubmission.js [hunk 10]
```diff
@@ -33,7 +37,10 @@ const submitPartial = () => {
     const fieldName = element.getAttribute('name');
     const value = getElementValue(element, fieldName);
     if (!data.has(fieldName)) {
-      if (typeof value === 'object') {
+      if (typeof value === 'object' && element.getAttribute('type') === 'file') {
+        data.append(fieldName, value);
+      }
```
Comment (lhalaa) [inline client/src/js/partialuserforms/partialsubmission.js]: Can we move `else if` to the same line as the closing bracket

File: client/src/js/partialuserforms/partialsubmission.js [hunk 11]
```diff
@@ -22,9 +22,13 @@ const getElementValue = (element, fieldName) => {
       return valueArray;
     }
     return '';
+  } else if (element.getAttribute('type') === 'file') {
+    return element.files[0];
```
Comment (lhalaa) [inline client/src/js/partialuserforms/partialsubmission.js:31]: Can you check if `element.files` is not empty, otherwise it will return undefined value

**Review by lhalaa — CHANGES_REQUESTED** [review]
Still not working

File: client/src/js/partialuserforms/partialsubmission.js (hunk 11 above)
Comment (Firesphere) [reply client/src/js/partialuserforms/partialsubmission.js:31]: The javascript itself doesn't seem to really care about it, but yeps :)

File: client/src/js/partialuserforms/partialsubmission.js (hunk 10 above)
Comment (Firesphere) [reply client/src/js/partialuserforms/partialsubmission.js]: Funny the linter didn't pick up on that!

File: src/controllers/PartialUserFormController.php [hunk 12]
```diff
@@ -45,54 +49,57 @@ public function partial(HTTPRequest $request)
...
+        if (!$request->getSession()->get(PartialSubmissionController::SESSION_KEY)) {
+            $request->getSession()->set(PartialSubmissionController::SESSION_KEY, $partial->ID);
+        }
 
-            // Set data record and load the form
-            $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
-            $this->setFailover($this->dataRecord);
+        // TODO: Recognize visitor with the password
+        // Set data record and load the form
+        $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
```
Comment (lhalaa) [inline src/controllers/PartialUserFormController.php:72]:
```suggestion
$record = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
$controller = parent::create($record);
$controller->doInit();
```
and replace `$this` with `$controller` seems to work. We don't have to set the failover manually.

Comment (lhalaa) [reply src/controllers/PartialUserFormController.php:72]: As discussed on Slack, this will be changed in the next PR.

**Review by lhalaa — APPROVED** [review]

**Pull request merged**